src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2288,9 +2288,15 @@ public override BoundNode VisitDefaultExpression(BoundDefaultExpression node)
             Debug.Assert(!IsConditionalState);
             if (this.State.Reachable)
             {
-                if ((object)node.Type != null && node.Type.IsReferenceType == true)
+                if ((object)node.Type != null && node.Type.IsReferenceType)
                 {
-                    this.State.ResultIsNotNull = false;
+                    bool isNotNull = node.Syntax.Kind() == SyntaxKind.DefaultExpression && !node.IsNullable;
```

Comment:
I didn't understand why we need to check the syntax kind.

Reply:
Because the expression `default(string)` should result in a warning by itself whereas the warning in `string s = default;` is from the assignment rather than the `default` expression.